Source/Scene/parseKTX2.js

```diff
@@ -0,0 +1,225 @@
+import arraySlice from "../Core/arraySlice.js";
+import defined from "../Core/defined.js";
+import PixelFormat from "../Core/PixelFormat.js";
+import RuntimeError from "../Core/RuntimeError.js";
+import VulkanConstants from "../Core/VulkanConstants.js";
+import PixelDatatype from "../Renderer/PixelDatatype.js";
+import { read } from "../ThirdParty/Workers/ktx-parse.modern.js";
+
+var faceOrder = [
+  "positiveX",
+  "negativeX",
+  "positiveY",
+  "negativeY",
+  "positiveZ",
+  "negativeZ",
+];
+
+// Flags
+var colorModelETC1S = 163;
+var colorModelUASTC = 166;
+
+/**
+ * Parses and transcodes KTX2 buffers to an appropriate target format.
+ *
+ * @param {Uint8Array} data Data representing one KTX2 texture.
+ * @param {Object} supportedTargetFormats Target formats available on the current system.
+ * @param {Object} transcoderModule basis_transcoder object.
+ *
+ * @private
+ */
+function parseKTX2(data, supportedTargetFormats, transcoderModule) {
+  if (!defined(supportedTargetFormats)) {
+    throw new RuntimeError(
+      "KTX2 Loader does not know which GPU formats are supported."
+    );
+  }
+
+  var header;
+  try {
+    header = read(data);
+  } catch (e) {
+    throw new RuntimeError("Invalid KTX2 file.");
+  }
+
+  if (header.layerCount !== 0) {
+    throw new RuntimeError("KTX2 Videos are not supported");
+  }
+
+  if (header.pixelDepth !== 0) {
+    throw new RuntimeError("KTX2 3D textures are unsupported.");
+  }
+
+  var dfd = header.dataFormatDescriptor[0];
+  var result = new Array(header.levelCount);
+
+  if (
+    header.vkFormat === 0x0 &&
+    (dfd.colorModel === colorModelETC1S || dfd.colorModel === colorModelUASTC)
+  ) {
+    // Compressed, initialize transcoder module
+    transcodeCompressed(
+      data,
+      header,
+      supportedTargetFormats,
+      transcoderModule,
+      result
+    );
+  } else {
+    parseUncompressed(header, result);
+  }
+
+  return result;
+}
+
+// Parser for uncompressed
+function parseUncompressed(header, result) {
+  var internalFormat = PixelFormat.RGBA;
+  var datatype;
+  if (header.vkFormat === VulkanConstants.VK_FORMAT_R8G8B8A8_UNORM) {
+    datatype = PixelDatatype.UNSIGNED_BYTE;
+  } else if (
+    header.vkFormat === VulkanConstants.VK_FORMAT_R16G16B16A16_SFLOAT
+  ) {
+    datatype = PixelDatatype.HALF_FLOAT;
+  } else if (
+    header.vkFormat === VulkanConstants.VK_FORMAT_R32G32B32A32_SFLOAT
+  ) {
+    datatype = PixelDatatype.FLOAT;
+  }
+
+  for (var i = 0; i < header.levels.length; ++i) {
+    var level = (result[i] = {});
+    var levelBuffer = arraySlice(header.levels[i].levelData);
```

Comment:
My thinking was based on this part in the [KTX-Parse docs](https://github.com/donmccurdy/ktx-parse#read) - 
> The container's mip levels and other binary data are pointers into the original file, not copies, so the original file should not be overwritten after reading.

I thought it'd be safer to not touch the original data when it gets passed to the `result` object,  but there might be a better way... the IBL sandcastle gets messed up without arraySlice here.

Reply:
Does our code actually modify the data? Also is it just the cube map that gets corrupted or does it happen to regular images as well?

On a related topic, whether it's possible to remove `arraySlice` or not, the underlying `ArrayBuffer` should still be added to `transferrableObjects` so that it's transferred rather than copied when it goes back to the main thread.

This can be done by adding `transferableObjects` to the `transcode` function in `transcodeKTX2` and passing it along to `parseKTX2`. `parseKTX2` should then push `ArrayBuffer`'s that are used by the result.

In general CesiumJS doesn't take advantage of transferring data back to the main thread as much as it should, though we do the reverse all the time.

```
function transcode(parameters, transferableObjects) {
  var ktx2Buffer = parameters.ktx2Buffer;
  var supportedTargetFormats = parameters.supportedTargetFormats;
  return parseKTX2(ktx2Buffer, supportedTargetFormats, transcoderModule, transferableObjects);
}
```

Reply:
Back to `arraySlice`...

I don't think we need it, the typed array constructors just needs to be given the proper byte offset into the `ArrayBuffer`.

`arraySlice` was hiding the problem because it created a new `ArrayBuffer` copy of the view. 

```js
      var faceByteOffset = levelBuffer.byteOffset + faceLength * j;

      var faceView;
      if (!defined(datatype) || PixelDatatype.sizeInBytes(datatype) === 1) {
        faceView = new Uint8Array(levelBuffer.buffer, faceByteOffset, faceLength);
      } else if (PixelDatatype.sizeInBytes(datatype) === 2) {
        faceView = new Uint16Array(levelBuffer.buffer, faceByteOffset, faceLength);
      } else {
        faceView = new Float32Array(levelBuffer.buffer, faceByteOffset, faceLength);
      }
```

Reply:
With both of these changes I think we can get to zero copies!

Reply:
I think the above fix only works for Uint8Array because `faceLength` should be the same in all three cases regardless of `header.typeSize`. With a couple small tweaks I think this is working now - 9273229 has no `arraySlice` in the worker and works for all three IBLs in the ktx2 sample folder.

Did we reach a conclusion about needing `arraySlice` in the main thread to avoid accessing the same glb for ktx2 + vertex data?

Reply:
> Did we reach a conclusion about needing `arraySlice` in the main thread to avoid accessing the same glb for ktx2 + vertex data?

Yes - I pushed the change in this commit: https://github.com/CesiumGS/cesium/pull/9513/commits/b33dc12d6b787fabf1034658907e7183e828e591. However I used the `Uint8Array` instead of `arraySlice` for performance reasons and because I knew the input and output would always be `Uint8Array`.